src/routes/events/containers/EventsPage.js

```diff
@@ -6,41 +6,85 @@ import Spinner from 'react-spinkit'
 
 import EventModel from 'modules/endpoint/models/EventModel'
 import EVENTS_ENDPOINT from 'modules/endpoint/endpoints/events'
+import LANGUAGES_ENDPOINT from 'modules/endpoint/endpoints/languages'
 import withFetcher from 'modules/endpoint/hocs/withFetcher'
-import withAvailableLanguageUpdater from 'modules/language/hocs/withAvailableLanguageUpdater'
-import EventDetail from './EventDetail'
+import EventDetail from '../components/EventDetail'
 import EventList from '../components/EventList'
+import { setLanguageChangeUrls } from 'modules/language/actions/setLanguageChangeUrls'
+import LanguageModel from 'modules/endpoint/models/LanguageModel'
 
 /**
  * Displays a list of events or a single event, matching the route /<location>/<language>/events*
  */
-class EventsPage extends React.Component {
+export class EventsPage extends React.Component {
   static propTypes = {
     events: PropTypes.arrayOf(PropTypes.instanceOf(EventModel)).isRequired,
     location: PropTypes.string.isRequired,
+    languages: PropTypes.arrayOf(PropTypes.instanceOf(LanguageModel)).isRequired,
     language: PropTypes.string.isRequired,
+    dispatchLanguageChangeUrls: PropTypes.func.isRequired,
     path: PropTypes.string
   }
 
   getPath () {
     return `/${this.props.location}/${this.props.language}/events`
   }
 
-  render () {
-    let events = this.props.events.map((event, index) => ({event: event, thumbnailPlaceholder: index % 3}))
+  mapLanguageToUrl = (language, id) =>
+    id ? `/${this.props.location}/${language}/events/${id}`
+      : `/${this.props.location}/${language}/events`
+
+  componentDidMount () {
+    // all events
+    let availableLanguages = {}
+
+    if (this.props.path && this.props.events) {
+      // specific event
+      const event = this.props.events.find(
+        (event) => event.id.toString() === this.props.path.replace('/', '')
+      )
+      if (event) availableLanguages = event.availableLanguages
+    }
+    this.props.dispatchLanguageChangeUrls(this.mapLanguageToUrl, this.props.languages, availableLanguages)
+  }
+
+  // we must not call dispatch in componentWillUpdate or componentDidUpdate
+  componentWillReceiveProps (nextProps) {
+    if (nextProps.events === this.props.events && nextProps.path === this.props.path) {
+      // no relevant prop changes
+      return
+    }
 
+    if (nextProps.path) {
```

Comment:
If you have only one id in the path and no hierarchy of paths, you can change the route definitoin in routes/index.js to '/events/:id'. Then you can access the id by state.router.params.id.

No more need for path parsing :)

Reply:
nice